File: 0124_binary_tree_maximum_path_sum/max_path_sum.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Time: O(n)
 * Space: O(1)
 */
func maxPathSum(root *TreeNode) int {
	res := []int{math.MinInt16}
	helper(root, res)
	return res[0]
}

func helper(node *TreeNode, res []int) int {
	if node == nil {
		return 0
	}
	left := helper(node.Left, res)
	right := helper(node.Right, res)
	res[0] = max(res[0], left+node.Val+right)
	return max(node.Val+max(left, right), 0)
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {

}

File: 0124_binary_tree_maximum_path_sum/max_path_sum.py
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.left = left
        self.right = right


def max_path_sum(root: TreeNode) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    res = [float('-inf')]
    helper(root, res)
    return res[0]


def helper(node: TreeNode, res: List[int]):
    if not node:
        return 0
    left = helper(node.left, res)
    right = helper(node.right, res)
    res[0] = max(res[0], left + node.val + right)
    return max(node.val + max(left, right), 0)
